x/upgrade/abci.go

```diff
@@ -9,17 +9,29 @@ import (
 )
 
 // BeginBlock will check if there is a scheduled plan and if it is ready to be executed.
+// If skip upgrade flag is set, it will skip and clear the upgrade plan
 // If it is ready, it will execute it if the handler is installed, and panic/abort otherwise.
 // If the plan is not ready, it will ensure the handler is not registered too early (and abort otherwise).
 //
-// The prupose is to ensure the binary is switch EXACTLY at the desired block, and to allow
+// The purpose is to ensure the binary is switched EXACTLY at the desired block, and to allow
 // a migration to be executed if needed upon this switch (migration defined in the new binary)
-func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock, skipUpgradeHeightArray []int64) {
+
 	plan, found := k.GetUpgradePlan(ctx)
 	if !found {
 		return
 	}
+
 	if plan.ShouldExecute(ctx) {
+		//To make sure clear upgrade is executed at the same block
+		if k.Contains(skipUpgradeHeightArray, ctx.BlockHeight()) {
+			// If skip upgrade has been set, we clear the upgrade plan
+			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %s: %s", plan.Name, plan.DueAt(), plan.Info)
```

Comment:
I would actually use `Height` instead of `DueAt` here, I think `DueAt` may return a timestamp as well.
Normally, `DueAt()` is nice, but since this skip is triggered by a flag with an exact block height, we should show this, so it lines up with the flag.

Reply:
Also, I note that there is no way to securely skip an upgrade that is triggered by time. This would need a wide range of skip flags to cover many possible heights. Once it fails upgrading, however, we know which block it is in and can skip, so this still works as an "abort upgrade after fail" approach, just not a pre-emptative upgrade.

I think handling this is out of scope of the PR, I just want to mention it, maybe Aaron wants to open a new issue to address that. (Possibly deprecating use of block time trigger)

Reply:
Yes, even I had this concern.Block height is a better option for any upgrade and all pervious upgrades mentioned a height on which the state will be dumped.
@aaronc